### PR TITLE
gpg pubkey keyhint import

### PR DESCRIPTION
**Pull request by mlandres**
[bsc#1184326](https://bugzilla.suse.com/show_bug.cgi?id=1184326)
- A few unrelated commits at the beginning
- Between download and signature verification, parse the master index for gpg keyhints, try to find/download those keys and import then into the public (untrusted) keyring. Their ids are then forwarded to the signatureVerificationWorkflow.
Todo:
- [x] adapt the signatureVerificationWorkflow to handle the additional keys and forward them in the callbacks
- [x] adapt/polish caching of downloaded keys

### CODE REVIEW

**Review by bzeller — CHANGES_REQUESTED** [review]

File: zypp/KeyRing.cc [hunk 1]
```diff
@@ -512,22 +543,30 @@ namespace zypp
...
       context_r.signatureIdTrusted( whichKeyring == trustedKeyRing() );
-      report->infoVerify( filedesc, foundKey, context );
+      report->infoVerify( filedesc, foundKey, keyContext );
       if ( verifyFile( file, signature, whichKeyring ) )
       {
 	context_r.fileValidated( true );
+	if ( context_r.signatureIdTrusted() && not buddies.empty() ) {
+	  // Check for buddy keys to be imported...
+	  MIL << "Validated with trusted key: importing buddy list..." << endl;
+	  for ( auto kd : buddies ) {
```
Comment (bzeller) [inline zypp/KeyRing.cc:553]: shouldn't that be const auto & ?  Otherwise you are copying all of them.

File: zypp/media/MediaCurl.cc [hunk 2]
```diff
@@ -346,7 +346,10 @@ void MediaCurl::setupEasy()
   {
 #if CURLVERSION_AT_LEAST(7,19,4)
     // restrict following of redirections from https to https only
-    SET_OPTION( CURLOPT_REDIR_PROTOCOLS, CURLPROTO_HTTPS );
+    if ( _url.getHost() == "download.opensuse.org" )
+      SET_OPTION( CURLOPT_REDIR_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS );
+    else
+      SET_OPTION( CURLOPT_REDIR_PROTOCOLS, CURLPROTO_HTTPS );
```
Comment (bzeller) [inline zypp/media/MediaCurl.cc:352]: Please also add this in  NetworkRequestPrivate::initialize

**Review by bzeller — CHANGES_REQUESTED** [review]

File: zypp/KeyRing.cc (hunk 1 above)
Comment (mlandres) [reply zypp/KeyRing.cc]: done

File: zypp/media/MediaCurl.cc (hunk 2 above)
Comment (mlandres) [reply zypp/media/MediaCurl.cc:352]: done

File: zypp/PublicKey.h [hunk 3]
```diff
@@ -220,9 +220,17 @@ namespace zypp
...
-    /** Whether \a id_r is the \ref id of the primary key or of a subkey. */
+    /** Whether \a id_r is the \ref id or \ref fingerprint of the primary key or of a subkey.
+     * As a convenience also allows to test the 8byte short ID e.g. rpm uses as version.
+     */
     bool providesKey( const std::string & id_r ) const;
 
+    /** The long id (64bit/16byte) or even better the fingerprint.
+     * The short Id (32bit/8byte) is not considered to be a safe identifier for a key.
+     */
+    static bool safeKeyId( const std::string & id_r )
```
[the remark this answers is not in the document]
Comment (mlandres) [reply zypp/PublicKey.h]: done

**Review by bzeller — APPROVED** [review]

**Pull request merged**